pkg/start/start.go

```diff
@@ -204,55 +192,34 @@ func (o *Options) run(ctx context.Context, controllerCtx *Context, lock *resourc
 		}()
 	}
 
-	exit := make(chan struct{})
-	exitClose := sync.Once{}
-
-	// TODO: when we switch to graceful lock shutdown, this can be
-	// moved back inside RunOrDie
-	// TODO: properly wire ctx here
-	go leaderelection.RunOrDie(context.TODO(), leaderelection.LeaderElectionConfig{
-		Lock:          lock,
-		LeaseDuration: leaseDuration,
-		RenewDeadline: renewDeadline,
-		RetryPeriod:   retryPeriod,
-		Callbacks: leaderelection.LeaderCallbacks{
-			OnStartedLeading: func(localCtx context.Context) {
-				controllerCtx.Start(runContext)
-				select {
-				case <-runContext.Done():
-					// WARNING: this is not completely safe until we have Kube 1.14 and ReleaseOnCancel
-					//   and client-go ContextCancelable, which allows us to block new API requests before
-					//   we step down. However, the CVO isn't that sensitive to races and can tolerate
-					//   brief overlap.
-					klog.Infof("Stepping down as leader")
-					// give the controllers some time to shut down
-					time.Sleep(100 * time.Millisecond)
-					// if we still hold the leader lease, clear the owner identity (other lease watchers
-					// still have to wait for expiration) like the new ReleaseOnCancel code will do.
-					if err := lock.Update(localCtx, resourcelock.LeaderElectionRecord{}); err == nil {
-						// if we successfully clear the owner identity, we can safely delete the record
-						if err := lock.Client.ConfigMaps(lock.ConfigMapMeta.Namespace).Delete(localCtx, lock.ConfigMapMeta.Name, metav1.DeleteOptions{}); err != nil {
-							klog.Warningf("Unable to step down cleanly: %v", err)
-						}
-					}
-					klog.Infof("Finished shutdown")
-					exitClose.Do(func() { close(exit) })
-				case <-localCtx.Done():
-					// we will exit in OnStoppedLeading
-				}
-			},
-			OnStoppedLeading: func() {
-				klog.Warning("leaderelection lost")
-				exitClose.Do(func() { close(exit) })
+	errorChannelCount++
+	go func() {
+		leaderelection.RunOrDie(runContext, leaderelection.LeaderElectionConfig{
+			Lock:            lock,
+			ReleaseOnCancel: true,
+			LeaseDuration:   leaseDuration,
+			RenewDeadline:   renewDeadline,
+			RetryPeriod:     retryPeriod,
+			Callbacks: leaderelection.LeaderCallbacks{
+				OnStartedLeading: func(ctx context.Context) {
+					controllerCtx.Start(ctx)
+				},
+				OnStoppedLeading: func() {
+					klog.Info("Stopped leading; shutting down.")
+					runCancel()
```

Comment:
I think you still need to exit, don't you?  How confident are you that this truly resets everything?

Reply:
> How confident are you that this truly resets everything?

If it doesn't, CI should turn it up, and we'll fix those bugs ;)

Reply:
This doesn't sound like what I described on slack.  If we lost lease, we exit immediately, no graceful step down.  When we have lost our lease we should not be running.

Reply:
> If we lost lease, we exit immediately...

`OnStoppedLeading` is not just "lost lease", it is also "gracefully released lease".  We could have logic here about checking `leaseContext.Err()` to guess about lost vs. released.  But `runCancel()` should immediately block all cluster-object-writing activity, so I think this is sufficient as it stands to keep from fighting the new leader.  For comparison, master gives [a full 5s grace period][1] after the cancel before forcing a hard exit.  If reducing my current 2m grace period to 5s would make you happy with this line, I'm happy to make that change.

[1]: https://github.com/openshift/cluster-version-operator/blob/16d3d8465b8e3f3c557becab28763d6cc18829a6/pkg/start/start.go#L156